pkg/frontend/querymiddleware/cardinality.go

```diff
@@ -0,0 +1,172 @@
+	// cardinalityEstimateBucketSize is the size of buckets that queries are bucketed into.
+	cardinalityEstimateBucketSize = 24 * time.Hour
+
+	// cacheErrorToleranceFraction is how much the estimate must deviate
+	// from the actually observed cardinality to update the cache.
+	cacheErrorToleranceFraction = 0.1
+)
+
+type cardinalityEstimationMetrics struct {
+	estimationError prometheus.Histogram
+}
+
+// cardinalityEstimation is a Handler that caches estimates for a query's
+// cardinality based on similar queries seen previously.
+type cardinalityEstimation struct {
+	cache  cache.Cache
+	next   Handler
+	logger log.Logger
+
+	cardinalityEstimationMetrics
+}
+
+func newCardinalityEstimationMiddleware(cache cache.Cache, logger log.Logger, registerer prometheus.Registerer) Middleware {
+	metrics := cardinalityEstimationMetrics{estimationError: promauto.With(registerer).NewHistogram(prometheus.HistogramOpts{
+		Name:    "cortex_frontend_cardinality_estimation_error",
+		Help:    "Difference between estimated and actual query cardinality",
+		Buckets: prometheus.ExponentialBuckets(10, 10, 6),
+	})}
+	return MiddlewareFunc(func(next Handler) Handler {
+		return &cardinalityEstimation{
+			cache:  cache,
+			next:   next,
+			logger: logger,
+
+			cardinalityEstimationMetrics: metrics,
+		}
+	})
+}
+
+// Do injects a cardinality estimate into the query hints (if available) and
+// caches the actual cardinality observed for this query.
+func (c *cardinalityEstimation) Do(ctx context.Context, request Request) (Response, error) {
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, c.logger, "cardinalityEstimation.Do")
+	defer spanLog.Finish()
+
+	tenants, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return c.next.Do(ctx, request)
+	}
+
+	k := generateCacheKey(tenant.JoinTenantIDs(tenants), request, cardinalityEstimateBucketSize)
+
+	var (
+		withCardinalityEstimate Request
+		estimatedCardinality    uint64
+	)
+	if estimate, ok := c.lookupCardinalityForKey(ctx, k); ok {
+		estimatedCardinality = estimate
+		withCardinalityEstimate = request.WithEstimatedCardinalityHint(estimate)
+	}
+
+	var res Response
+	if withCardinalityEstimate != nil {
+		res, err = c.next.Do(ctx, withCardinalityEstimate)
+	} else {
+		res, err = c.next.Do(ctx, request)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	statistics := stats.FromContext(ctx)
+	if statistics == nil {
+		return res, nil
+	}
+
+	actualCardinality := statistics.GetFetchedSeriesCount()
+	if !isCardinalitySimilar(actualCardinality, estimatedCardinality) {
+		c.storeCardinalityForKey(ctx, k, actualCardinality)
+	}
+
+	c.maintainStatistics(estimatedCardinality, actualCardinality, statistics, spanLog)
+
+	return res, nil
+}
+
+// lookupCardinalityForKey fetches a cardinality estimate for the given key from
+// the results cache.
+func (c *cardinalityEstimation) lookupCardinalityForKey(ctx context.Context, key string) (uint64, bool) {
+	if c.cache == nil {
+		return 0, false
+	}
+	cacheKey := cacheHashKey(key)
```

Comment:
I used this because of this comment
https://github.com/grafana/mimir/blob/4ee079a33a6057e5431d2202b053b6ef8780b9ce/pkg/frontend/querymiddleware/split_and_cache.go#L619

Reply:
Right. For the cardinality middleware, I think you should just hash the `Query()` portion of the cache key. Everything else should not be hashed.